File: watch-app/entry/src/main/js/MainAbility/common/alarmHash.js
// Canonical hash of the local alarm set. Mirrors `AlarmHash.kt` on the
// phone byte-for-byte. The phone polls the watch with `sync_check`; we
// reply with `sync_hash` carrying this 8-char hex. If the phone's
// computed hash matches, it skips the full alarm-by-alarm push.
//
// Algorithm:
//   1. Sort alarms by id ascending
//   2. Render each as a pipe-delimited line:
//        id|label|hour|minute|daysOfWeek|enabled|audioUri|
//        isVibrationOnly|snoozeMinutes|updatedAtEpoch|
//        relativeMinutes|selfDestruct
//      - booleans: '1' / '0'
//      - null audioUri / undefined relativeMinutes: empty string
//   3. Join with '\n' (newline after EVERY line, including last)
//   4. hash = Java String.hashCode → unsigned → 8-char lowercase hex
//
// Java String.hashCode contract: `h = h * 31 + char` iterated over UTF-16
// code units, with overflow wrapping inside 32-bit signed int. JS
// expresses that with `((h << 5) - h + c) | 0` — `h * 31 == (h << 5) - h`
// and `| 0` collapses to 32-bit signed semantics on every step.
//
// Any divergence from the Kotlin canonicalize() function makes hash
// compare always-fail and the optimization becomes dead overhead. If
// either side changes the format, update BOTH in the same commit.

function boolFlag(v) {
    return v ? '1' : '0';
}

function strOrEmpty(v) {
    if (v === null || v === undefined) return '';
    return '' + v;
}

function sortById(alarms) {
    var sorted = alarms.slice();
    sorted.sort(function (a, b) {
        if (a.id < b.id) return -1;
        if (a.id > b.id) return 1;
        return 0;
    });
    return sorted;
}

function canonicalize(alarms) {
    var sorted = sortById(alarms);
    var out = '';
    for (var i = 0; i < sorted.length; i++) {
        var a = sorted[i];
        out += a.id + '|' +
            strOrEmpty(a.label) + '|' +
            a.hour + '|' +
            a.minute + '|' +
            (a.daysOfWeek || 0) + '|' +
            boolFlag(a.enabled) + '|' +
            strOrEmpty(a.audioUri) + '|' +
            boolFlag(a.isVibrationOnly) + '|' +
            (a.snoozeMinutes || 0) + '|' +
            (a.updatedAtEpoch || 0) + '|' +
            strOrEmpty(a.relativeMinutes) + '|' +
            boolFlag(a.selfDestruct) +
            '\n';
    }
    return out;
}

function javaHashCode(s) {
    var h = 0;
    for (var i = 0; i < s.length; i++) {
        h = ((h << 5) - h + s.charCodeAt(i)) | 0;
    }
    return h;
}

function toUnsignedHex8(h32) {
    // Convert 32-bit signed to unsigned, then to lowercase hex padded to 8.
    var u = h32 >>> 0;
    var hex = u.toString(16);
    while (hex.length < 8) hex = '0' + hex;
    return hex;
}

export default {
    compute: function (alarms) {
        var canonical = canonicalize(alarms || []);
        return toUnsignedHex8(javaHashCode(canonical));
    },
};

File: android-app/app/src/main/java/com/kirkouski/gtalarm/data/sync/AlarmHash.kt
package com.kirkouski.gtalarm.data.sync

import com.kirkouski.gtalarm.domain.Alarm

/**
 * Deterministic 8-char hex hash of an alarm set. Used by the sync-check
 * round-trip (phone → `sync_check` → watch → `sync_hash` → phone) to skip
 * a full alarm-by-alarm push when the watch already mirrors the phone's
 * state.
 *
 * **Algorithm:** canonicalize the alarm list to a stable text representation,
 * then apply Java `String.hashCode()`. The hashCode algorithm is spec'd as
 * `h * 31 + c` iterated over UTF-16 code units — trivially mirrored in JS
 * on the watch side (`((h << 5) - h + s.charCodeAt(i)) | 0`).
 *
 * **Canonical form:** alarms sorted by id ascending, each rendered as a
 * pipe-delimited line `id|hour|minute|daysOfWeek|enabled|audioUri|
 * isVibrationOnly|snoozeMinutes|updatedAtEpoch|relativeMinutes|selfDestruct`
 * with `\n` separator. Booleans render as `1`/`0`; null audioUri / null
 * relativeMinutes render as empty string between the surrounding pipes.
 *
 * Match the JS implementation in `watch-app/.../common/alarmHash.js` BYTE
 * FOR BYTE — any divergence makes the hash compare always-fail and the
 * check becomes useless overhead.
 *
 * **Intentionally NOT included** in the canonical form:
 *   - `label` — not part of the watch wire format at all; the alarm label
 *     is phone-only (shown on the phone ring screen). The watch never
 *     receives it, so including it here would force a permanent mismatch.
 *   - `backgroundImageUri` — device-local `content://` path that doesn't
 *     round-trip between phone and watch (watch never has the same URI).
 *     Reconciliation of the watch image happens via file-transfer of the
 *     BGRA `.bin` blob, orthogonal to this hash. Code-review pass 1
 *     (2026-05-13) confirmed including it broke the AlreadyInSync
 *     short-circuit on every sync.
 */
object AlarmHash {
    fun compute(alarms: List<Alarm>): String {
        val canonical = canonicalize(alarms)
        return canonical.hashCode().toUInt().toString(16).padStart(HEX_WIDTH, '0')
    }

    private fun canonicalize(alarms: List<Alarm>): String {
        val sb = StringBuilder()
        alarms.sortedBy { it.id }.forEach { a ->
            sb.append(a.id).append('|')
                .append(a.hour).append('|')
                .append(a.minute).append('|')
                .append(a.daysOfWeek).append('|')
                .append(if (a.enabled) '1' else '0').append('|')
                .append(a.audioUri.orEmpty()).append('|')
                .append(if (a.isVibrationOnly) '1' else '0').append('|')
                .append(a.snoozeMinutes).append('|')
                .append(a.updatedAtEpoch).append('|')
                .append(a.relativeMinutes?.toString().orEmpty()).append('|')
                .append(if (a.selfDestruct) '1' else '0')
                .append('\n')
        }
        return sb.toString()
    }

    private const val HEX_WIDTH = 8
}
